[analyzer] Eliminate unique release point assertion

NagyDonat opened the pull request:
MallocChecker.cpp has a complex heuristic that supresses reports where the memory release happens during the release of a reference-counted object (to suppress a significant amount of false positives).

Previously this logic asserted that there is at most one release point corresponding to a symbol, but it turns out that there is a rare corner case where the symbol can be released, forgotten and then released again. This commit removes that assertion to avoid the crash. (As this issue just affects a bug suppression heuristic, I didn't want to dig deeper and modify the way the state of the symbol is changed.)

Fixes #149754

NagyDonat committed:
[analyzer] Eliminate unique release point assertion

MallocChecker.cpp has a complex heuristic that supresses reports where
the memory release happens during the release of a reference-counted
object (to suppress a significant amount of false positives).

Previously this logic asserted that there is at most one release point
corresponding to a symbol, but it turns out that there is a rare corner
case where the symbol can be released, forgotten and then released
again. This commit removes that assertion to avoid the crash. (As this
issue just affects a bug suppression heuristic, I didn't want to dig
deeper in the state changes.)

Fixes #149754

NagyDonat commented:
On Windows `malloc.c` started to fail with
```
2025-07-23T15:54:30.2721710Z # RUN: at line 1
2025-07-23T15:54:30.2724937Z c:\_work\llvm-project\llvm-project\build\bin\clang.exe -cc1 -internal-isystem C:\_work\llvm-project\llvm-project\build\lib\clang\22\include -nostdsysteminc -analyze -setup-static-analyzer -Wno-strict-prototypes -Wno-error=implicit-int -verify C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c    -analyzer-checker=core    -analyzer-checker=alpha.deadcode.UnreachableCode    -analyzer-checker=alpha.core.CastSize    -analyzer-checker=unix    -analyzer-checker=debug.ExprInspection    -analyzer-checker=optin.taint.TaintPropagation    -analyzer-checker=optin.taint.TaintedAlloc
2025-07-23T15:54:30.2731658Z # executed command: 'c:\_work\llvm-project\llvm-project\build\bin\clang.exe' -cc1 -internal-isystem 'C:\_work\llvm-project\llvm-project\build\lib\clang\22\include' -nostdsysteminc -analyze -setup-static-analyzer -Wno-strict-prototypes -Wno-error=implicit-int -verify 'C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c' -analyzer-checker=core -analyzer-checker=alpha.deadcode.UnreachableCode -analyzer-checker=alpha.core.CastSize -analyzer-checker=unix -analyzer-checker=debug.ExprInspection -analyzer-checker=optin.taint.TaintPropagation -analyzer-checker=optin.taint.TaintedAlloc
2025-07-23T15:54:30.2735938Z # .---command stderr------------
2025-07-23T15:54:30.2736504Z # | error: 'expected-warning' diagnostics expected but not seen: 
2025-07-23T15:54:30.2737888Z # |   File C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c Line 1956 (directive at C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c:1957): {{^conj_\$[[:digit:]]+{int, LC1, S[[:digit:]]+, #1}}
2025-07-23T15:54:30.2739908Z # |   File C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c Line 1959 (directive at C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c:1960): {{^conj_\$[[:digit:]]+{int, LC1, S[[:digit:]]+, #1}}
2025-07-23T15:54:30.2741155Z # | error: 'expected-warning' diagnostics seen but not expected: 
2025-07-23T15:54:30.2742033Z # |   File C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c Line 1956: conj_$4{int, LC2, S54828, #1} [debug.ExprInspection]
2025-07-23T15:54:30.2743218Z # |   File C:\_work\llvm-project\llvm-project\clang\test\Analysis\malloc.c Line 1959: conj_$4{int, LC2, S54828, #1} [debug.ExprInspection]
2025-07-23T15:54:30.2744021Z # | 4 errors generated.
2025-07-23T15:54:30.2744346Z # `-----------------------------
2025-07-23T15:54:30.2744718Z # error: command failed with exit status: 1
```
I'll investigate this on the upcoming days.

pskrgag reviewed (approved):
LGTM! Not sure why windows CI is unhappy, but the change makes total sense to me

NagyDonat committed:
Don't check location context id value in test

The ID of a location context is a completely arbitrary number
(`LocationContextManager` issues consecutive numbers as the IDs of the
location contexts that it constructs) which is irrelevant for that
particular test, so use a wildcard for matching it (instead of a
concrete number).

This check made the test break on a Windows buildbot, which I cannot
investigate (because I don't have access to Windows), so I'm
sidestepping the issue with this commit.

NagyDonat commented:
I created a commit to fix the issue that was detected by the Windows buildbot. I'm planning to merge this commit if the Windows buildbot succeeds with the new revision (and there are no other suggestions).
